src/fastpath/fastpath.c

```diff
@@ -6,15 +6,313 @@
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+
+                /* Slowpath FPU saving case. We check that dest either isn't currently using the FPU
+                 * so that we do not have to save its state for migration (lazy FPU) or that it is the
+                 * the thread that is currently using the FPU, but the CPU is the same as that of the
+                 * scheduling context so that tcb migration is not needed. Not exactly the same as the
+                 * slowpath as this saves FPU state in the second case but there should be no reason
+                 * to touch the FPU if migration is not required */
+#ifdef ENABLE_SMP_SUPPORT
+#ifdef CONFIG_HAVE_FPU
+                if (nativeThreadUsingFPU(dest)) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+#endif
+#endif
+            } else {
+                sc = dest->tcbSchedContext;
+            }
+
+            /* Signal to higher prio thread is NOT fastpathed. Not sure if this handles idle thread */
+            if (NODE_STATE_ON_CORE(ksCurThread, sc->scCore)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Simplified schedContext_resume that does not change state and reverts to the
+            * slowpath in cases where the SC does not have sufficient budget, as this case
+            * adds extra scheduler logic. Normally, this is done after the sched_context donate
+            * but after tweaking it, I don't immediately see anything executed in schedContext_donate
+            * that will affect the conditions of this check */
+
+            if (sc->scRefillMax > 0 && !thread_state_get_tcbInReleaseQueue(dest->tcbState)) {
+                if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+                schedulable = true;
+            }
+
+            if (ksCurDomain != dest->tcbDomain SMP_COND_STATEMENT( || sc->scCore != getCurrentCPUIndex())) {
+                crossnode = true;
+            }
+
+            /*  Point of no return */
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            tcb_queue_t queue;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+
+            queue = ep_ptr_get_queue(ep_ptr);
+            queue = tcbEPDequeue(dest, queue);
+            ep_ptr_set_queue(ep_ptr, queue);
+
+            if (!queue.head) {
+                endpoint_ptr_set_state(ep_ptr, EPState_Idle);
+            }
+
+            reply_t *reply = REPLY_PTR(thread_state_get_replyObject(dest->tcbState));
+            if (reply != NULL) {
+                reply_unlink(reply, dest);
+            }
```

Comment:
Is this reply object logic a recent addition to this PR? If the dest thread is `ThreadState_BlockedOnReceive` then it can't be blocked on a reply object. (There should be no way for signal to wake up a thread blocked on a reply object)

Reply:
I think that's always been there as `reply_unlink()` is part of `cancel_ipc()` for `ThreadState_BlockedOnReceive`. I think it's for the case where the target thread is blocked after doing a `seL4_Recv()` and is waiting for an IPC to come in, in which case you want to cancel this recv and allow the signal to happen instead.

Reply:
Oh right, this is the link between the receiving thread and it's reply object to track scheduling context donation across a call.  I missed that this was part of the cancelIPC operation for `ThreadState_BlockedOnReceive`